components/codec/buffer.rs

```diff
@@ -0,0 +1,435 @@
+    }
+}
+
+impl<'a> BufferReader for &'a [u8] {
+    fn bytes(&self) -> &[u8] {
+        self
+    }
+
+    fn advance(&mut self, count: usize) {
+        *self = &self[count..]
+    }
+}
+
+impl<'a, T: BufferReader + ?Sized> BufferReader for &'a mut T {
+    fn bytes(&self) -> &[u8] {
+        (**self).bytes()
+    }
+
+    fn advance(&mut self, count: usize) {
+        (**self).advance(count)
+    }
+}
+
+impl<T: BufferReader + ?Sized> BufferReader for Box<T> {
+    fn bytes(&self) -> &[u8] {
+        (**self).bytes()
+    }
+
+    fn advance(&mut self, count: usize) {
+        (**self).advance(count)
+    }
+}
+
+/// A trait to provide sequential write over a fixed size
+/// or dynamic size memory buffer.
+///
+/// The memory buffer can be `std::io::Cursor<AsRef<[u8]>>` or `&mut [u8]`,
+/// which is fixed sized, or `Vec<u8>`, which is dynamically sized.
+pub trait BufferWriter {
+    /// Returns a mutable slice starting at current position.
+    ///
+    /// The caller may hint the underlying buffer to grow according to `size`
+    /// if the underlying buffer is dynamically sized (i.e. is capable to grow).
+    ///
+    /// The size of the returned slice may be less than `size` given. For example,
+    /// when underlying buffer is fixed sized and there is no enough space any more.
+    ///
+    /// The returned mutable slice is for writing only and should be never used for
+    /// reading since it might contain uninitialized memory when underlying buffer
+    /// is dynamically sized. For this reason, this function is marked `unsafe`.
+    unsafe fn bytes_mut(&mut self, size: usize) -> &mut [u8];
+
+    /// Advances the position of internal cursor for a previous write.
+    ///
+    /// The caller should ensure that advanced positions have been all written
+    /// previously. If the cursor is moved beyond actually written data, it will
+    /// leave uninitialized memory. For this reason, this function is marked
+    /// `unsafe`.
+    ///
+    /// # Panics
+    ///
+    /// This function may panic in some implementors when remaining space
+    /// is not large enough to advance.
+    unsafe fn advance_mut(&mut self, count: usize);
+}
+
+impl<T: AsMut<[u8]>> BufferWriter for ::std::io::Cursor<T> {
+    unsafe fn bytes_mut(&mut self, _size: usize) -> &mut [u8] {
+        // `size` is ignored since this buffer is not capable to grow.
+        let pos = self.position() as usize;
+        let slice = self.get_mut().as_mut();
+        if pos >= slice.len() {
+            return &mut [];
+        }
+        &mut slice[pos..]
+    }
+
+    unsafe fn advance_mut(&mut self, count: usize) {
+        let mut pos = self.position();
+        pos += count as u64;
+        self.set_position(pos);
+    }
+}
+
+impl<'a> BufferWriter for &'a mut [u8] {
+    unsafe fn bytes_mut(&mut self, _size: usize) -> &mut [u8] {
+        self
+    }
+
+    unsafe fn advance_mut(&mut self, count: usize) {
+        let original_self = ::std::mem::replace(self, &mut []);
```

Comment:
Can we just `*self = &mut self[count..]`?

Reply:
cannot. rust reports borrow issue and seems that using mem::replace is the only solution, which is used [in std](https://doc.rust-lang.org/src/core/slice/mod.rs.html#3846) as well